src/api/unified/symbol_manager.cpp

```diff
@@ -8,145 +8,128 @@
+
+#include <cmath>
+#include <string>
+#include <type_traits>
+
+
+#ifndef WIN_OS
+#include <dlfcn.h>
+#else
+#include <Windows.h>
+#endif
+
+using common::loadLibrary;
+using common::unloadLibrary;
 
+using std::extent;
 using std::string;
-using std::replace;
 
 namespace unified
 {
 
-static const char* LIB_AF_BKND_NAME[NUM_BACKENDS] = {"cpu", "cuda", "opencl"};
 #if defined(OS_WIN)
-static const char* LIB_AF_BKND_PREFIX = "af";
+static const char* LIB_AF_BKND_PREFIX = "";
 static const char* LIB_AF_BKND_SUFFIX = ".dll";
+#define PATH_SEPARATOR "\\"
 #define RTLD_LAZY 0
 #else
+
 #if defined(__APPLE__)
-#define SO_SUFFIX_HELPER(VER) "." #VER ".dylib"
+#  define SO_SUFFIX_HELPER(VER) "." #VER ".dylib"
 #else
-#define SO_SUFFIX_HELPER(VER) ".so." #VER
-#endif // APPLE
-static const char* LIB_AF_BKND_PREFIX = "libaf";
+#  define SO_SUFFIX_HELPER(VER) ".so." #VER
+#endif
+   static const char* LIB_AF_BKND_PREFIX = "lib";
+#  define PATH_SEPARATOR "/"
 
-#define GET_SO_SUFFIX(VER) SO_SUFFIX_HELPER(VER)
-static const char* LIB_AF_BKND_SUFFIX = GET_SO_SUFFIX(AF_VERSION_MAJOR);
+#  define GET_SO_SUFFIX(VER) SO_SUFFIX_HELPER(VER)
+   static const char* LIB_AF_BKND_SUFFIX = GET_SO_SUFFIX(AF_VERSION_MAJOR);
 #endif
 
-static const char* LIB_AF_ENVARS[NUM_ENV_VARS] = {"AF_PATH", "AF_BUILD_PATH"};
-static const char* LIB_AF_RPATHS[NUM_ENV_VARS] = {"/lib/", "/src/backend/"};
-static const bool LIB_AF_RPATH_SUFFIX[NUM_ENV_VARS] = {false, true};
+string getBkndLibName(const af_backend backend) {
+    string ret;
+    switch (backend) {
+        case AF_BACKEND_CUDA: ret = string(LIB_AF_BKND_PREFIX) + "afcuda" + LIB_AF_BKND_SUFFIX; break;
+        case AF_BACKEND_OPENCL: ret = string(LIB_AF_BKND_PREFIX) + "afopencl" + LIB_AF_BKND_SUFFIX; break;
+        case AF_BACKEND_CPU: ret = string(LIB_AF_BKND_PREFIX) + "afcpu" + LIB_AF_BKND_SUFFIX; break;
+        default: assert(1!=1 && "Invalid backend");
+    }
+    return ret;
+}
+string getBackendDirectoryName(const af_backend backend) {
+    string ret;
+    switch (backend) {
+        case AF_BACKEND_CUDA: ret = "cuda"; break;
+        case AF_BACKEND_OPENCL: ret = "opencl"; break;
+        case AF_BACKEND_CPU: ret = "cpu"; break;
+        default: assert(1!=1 && "Invalid backend");
+      }
+    return ret;
+}
 
-inline string getBkndLibName(const int backend_index)
-{
-    int i = backend_index >=0 && backend_index<NUM_BACKENDS ? backend_index : 0;
-    return string(LIB_AF_BKND_PREFIX) + LIB_AF_BKND_NAME[i] + LIB_AF_BKND_SUFFIX;
+string join_path(string first) {
+    return first;
+}
+
+template<typename... ARGS>
+string join_path(string first, ARGS... args) {
+    if(first.empty()) { return join_path(args...); }
+    else              { return first + PATH_SEPARATOR + join_path(args...); }
 }
 
 /*flag parameter is not used on windows platform */
-LibHandle openDynLibrary(const int bknd_idx, int flag=RTLD_LAZY)
+LibHandle openDynLibrary(const af_backend bknd_idx, int flag=RTLD_LAZY)
```

Comment:
you might want to change the variable `bknd_idx` name to something else, now that it is an enum and not an integer.